### PR TITLE
GraphQL 0.8.0 -> 0.8.2

### PR DESCRIPTION
**Pull request by intellix**
Please fill in this template.

- [x] Prefer to make your PR against the `types-2.0` branch.
- [x] Test the change in your own code.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `npm run lint -- package-name` if a `tslint.json` is present.

If changing an existing definition:
- [x] Provide a URL to  documentation or source code which provides context for the suggested changes: <https://github.com/graphql/graphql-js/compare/v0.8.0...v0.8.2?diff=split&name=v0.8.2>
- [x] Increase the version number in the header if appropriate.

@TonyPythoneer @calebmer

### CODE REVIEW

**Comment by intellix** [comment]
Future TODO:

Try and keep the same format with exports. Currently there are a few instances of `export *` of something when upstream has a list of exports. I think this makes it harder to update this package as it's confusing where stuff is.

From
```
declare module "graphql/language/index" {
    export * from 'graphql/language/ast';
}
```

To:
```
declare module "graphql/language/index" {
  export {
    ...
    InputObjectTypeDefinition,
    TypeExtensionDefinition,
    DirectiveDefinition,
  } from '/graphql/language/ast';
}
```

I don't think it's a showstopper, it would just make maintenance easier in future

**Comment by intellix** [comment]
It seems that GraphQLError isn't quite correct here: 
declared class: https://github.com/graphql/graphql-js/blob/master/src/error/GraphQLError.js#L21
exported fn: https://github.com/graphql/graphql-js/blob/master/src/error/GraphQLError.js#L72

I'm using this PR against apollo-client with the latest types from here and getting errors with this stuff: https://github.com/apollostack/apollo-client/blob/master/test/ObservableQuery.ts#L473
```
test/ObservableQuery.ts(473,45): error TS2345: Argument of type '{ request: { query: any; variables: { id: number; }; }; result: { errors: { name: string; message...' is not assignable to parameter of type
 'MockedResponse'.
  Types of property 'result' are incompatible.
    Type '{ errors: { name: string; message: string; }[]; }' is not assignable to type 'ExecutionResult'.
      Types of property 'errors' are incompatible.
        Type '{ name: string; message: string; }[]' is not assignable to type 'GraphQLError[]'.
          Type '{ name: string; message: string; }' is not assignable to type 'GraphQLError'.
            Property 'locations' is missing in type '{ name: string; message: string; }'.
```

To be honest I'm not quite sure what to do here, if anyone has a suggestion

File: graphql/index.d.ts [hunk 1]
```diff
@@ -2781,7 +2794,40 @@ declare module "graphql/utilities/typeFromAST" {
...
+     * of the current field and type definitions at any point in a GraphQL document
+     * AST during a recursive descent by calling `enter(node)` and `leave(node)`.
+     */
+    class TypeInfo {
+        constructor(
+            schema: GraphQLSchema,
+            // NOTE: this experimental optional second parameter is only needed in order
+            // to support non-spec-compliant codebases. You should never need to use it.
+            // It may disappear in the future.
+            getFieldDefFn: Function
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: It should be `getFieldDefFn: getFieldDef`.

Solution

getFieldDef:
```
export interface getFieldDef {
  (
    schema: GraphQLSchema,
    parentType: GraphQLType,
    fieldNode: FieldNode
  ): GraphQLField
}
```

Although getFieldDef is a function and doesn't export, I think it needs to build a type in module.

File: graphql/index.d.ts [hunk 2]
```diff
@@ -2007,7 +2020,7 @@ declare module "graphql/execution/values" {
      * definitions and list of argument AST nodes.
      */
     function getArgumentValues(
-        def: GraphQLField | GraphQLDirective,
+        def: GraphQLField<any, any> | GraphQLDirective,
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 3]
```diff
@@ -1983,7 +1996,7 @@ declare module "graphql/execution/execute" {
      * and returns it as the result, or if it's a function, returns the result
      * of calling that function while passing along args and context.
      */
-    export const defaultFieldResolver: GraphQLFieldResolver<any>;
+    export const defaultFieldResolver: GraphQLFieldResolver<any, any>;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 4]
```diff
@@ -1944,7 +1957,7 @@ declare module "graphql/execution/execute" {
      * non-empty array if an error occurred.
      */
     export interface ExecutionResult {
-        data: {[key: string]: any};
+        data?: {[key: string]: any};
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 5]
```diff
@@ -1889,7 +1902,7 @@ declare module "graphql/validation/validate" {
 
         getInputType(): GraphQLInputType;
 
-        getFieldDef(): GraphQLField;
+        getFieldDef(): GraphQLField<any, any>;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 6]
```diff
@@ -1680,9 +1693,9 @@ declare module "graphql/type/introspection" {
      * Note that these are GraphQLField and not GraphQLFieldConfig,
      * so the format for args is different.
      */
-    const SchemaMetaFieldDef: GraphQLField;
-    const TypeMetaFieldDef: GraphQLField;
-    const TypeNameMetaFieldDef: GraphQLField;
+    const SchemaMetaFieldDef: GraphQLField<any, any>;
+    const TypeMetaFieldDef: GraphQLField<any, any>;
+    const TypeNameMetaFieldDef: GraphQLField<any, any>;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 7]
```diff
@@ -1380,24 +1393,24 @@ declare module "graphql/type/definition" {
     class GraphQLUnionType {
         name: string;
         description: string;
-        resolveType: GraphQLTypeResolver;
+        resolveType: GraphQLTypeResolver<any, any>;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 8]
```diff
@@ -1380,24 +1393,24 @@ declare module "graphql/type/definition" {
...
 
-        constructor(config: GraphQLUnionTypeConfig);
+        constructor(config: GraphQLUnionTypeConfig<any, any>);
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 9]
```diff
@@ -1308,8 +1321,8 @@ declare module "graphql/type/definition" {
         description?: string;
     }
 
-    export interface GraphQLFieldMap {
-        [fieldName: string]: GraphQLField;
+    export interface GraphQLFieldMap<TSource, TContext> {
+        [fieldName: string]: GraphQLField<TSource, TContext>;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 10]
```diff
@@ -1287,18 +1300,18 @@ declare module "graphql/type/definition" {
...
     }
 
-    export interface GraphQLFieldConfigMap<TSource> {
-        [fieldName: string]: GraphQLFieldConfig<TSource>;
+    export interface GraphQLFieldConfigMap<TSource, TContext> {
+        [fieldName: string]: GraphQLFieldConfig<TSource, TContext>;
     }
 
-    export interface GraphQLField {
+    export interface GraphQLField<TSource, TContext> {
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 11]
```diff
@@ -1287,18 +1300,18 @@ declare module "graphql/type/definition" {
         description?: string;
     }
 
-    export interface GraphQLFieldConfigMap<TSource> {
-        [fieldName: string]: GraphQLFieldConfig<TSource>;
+    export interface GraphQLFieldConfigMap<TSource, TContext> {
+        [fieldName: string]: GraphQLFieldConfig<TSource, TContext>;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 12]
```diff
@@ -1269,10 +1282,10 @@ declare module "graphql/type/definition" {
 
     export type ResponsePath = { prev: ResponsePath, key: string | number } | void;
 
-    export interface GraphQLFieldConfig<TSource> {
+    export interface GraphQLFieldConfig<TSource, TContext> {
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 13]
```diff
@@ -1217,40 +1230,40 @@ declare module "graphql/type/definition" {
...
     ) => boolean;
 
-    export type GraphQLFieldResolver<TSource> = (
+    export type GraphQLFieldResolver<TSource, TContext> = (
         source: TSource,
         args: { [argName: string]: any },
-        context: any,
+        context: TContext,
         info: GraphQLResolveInfo
     ) => any;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 14]
```diff
@@ -1217,40 +1230,40 @@ declare module "graphql/type/definition" {
...
     ) => GraphQLObjectType;
 
-    export type GraphQLIsTypeOfFn = (
-        source: any,
-        context: any,
+    export type GraphQLIsTypeOfFn<TSource, TContext> = (
+        source: TSource,
+        context: TContext,
         info: GraphQLResolveInfo
     ) => boolean;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 15]
```diff
@@ -1217,40 +1230,40 @@ declare module "graphql/type/definition" {
...
-        getFields(): GraphQLFieldMap;
+        constructor(config: GraphQLObjectTypeConfig<any, any>);
+        getFields(): GraphQLFieldMap<any, any>;
         getInterfaces(): Array<GraphQLInterfaceType>;
         toString(): string;
     }
 
     //
 
-    export interface GraphQLObjectTypeConfig<TSource> {
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts [hunk 16]
```diff
@@ -1100,6 +1111,8 @@ declare module "graphql/type/definition" {
 
     export function isAbstractType(type: GraphQLType): type is GraphQLAbstractType;
 
+    export function assertAbstractType(type: GraphQLType): GraphQLAbstractType;
```
Comment (TonyPythoneer) [inline graphql/index.d.ts]: OK.

File: graphql/index.d.ts (hunk 1 above)
Comment (TonyPythoneer) [inline graphql/index.d.ts]: @intellix Here need to fix.

**Comment by TonyPythoneer** [comment]
@intellix  Okay, it's great. I have reviewed it.
It has one line which needs to fix. :smile:

File: graphql/index.d.ts (hunk 1 above)
Comment (intellix) [inline graphql/index.d.ts]: fixed

**Comment by intellix** [comment]
Fixed that one thing. What do you think to GraphQLError being incorrectly exported? https://github.com/DefinitelyTyped/DefinitelyTyped/pull/12767#issuecomment-261411364

Add a constructor?
```
class GraphQLError extends Error {

        /**
         * A message describing the Error for debugging purposes.
         *
         * Enumerable, and appears in the result of JSON.stringify().
         */
        message: string;

        /**
         * An array of { line, column } locations within the source GraphQL document
         * which correspond to this error.
         *
         * Errors during validation often contain multiple locations, for example to
         * point out two things with the same name. Errors during execution include a
         * single location, the field which produced the error.
         *
         * Enumerable, and appears in the result of JSON.stringify().
         */
        locations: Array<{ line: number, column: number }> | void;

        /**
         * An array describing the JSON-path into the execution response which
         * corresponds to this error. Only included for errors during execution.
         *
         * Enumerable, and appears in the result of JSON.stringify().
         */
        path: Array<string | number> | void;

        /**
         * An array of GraphQL AST Nodes corresponding to this error.
         */
        nodes: Array<ASTNode> | void;

        /**
         * The source GraphQL document corresponding to this error.
         */
        source: Source | void;

        /**
         * An array of character offsets within the source GraphQL document
         * which correspond to this error.
         */
        positions: Array<number> | void;

        /**
         * The original error thrown from a field resolver during execution.
         */
        originalError: Error;

        constructor(
            message: string,
            nodes?: Array<any>,
            source?: Source,
            positions?: Array<number>,
            path?: Array<string | number>,
            originalError?: Error,
        );
    }
```

**Comment by TonyPythoneer** [comment]
@intellix I think we should open an issue to request to change the naming.

Personally, I don't like t what GraphQL team names class and function as `GraphQLError` because it maybe will happen `Duplicate identifier` in the program. Therefore, I hope pending this change and let this PR gets merged. Until it has got solved by GraphQL team, you can send a new PR to fix this. How do you think?

**Comment by intellix** [comment]
These properties of the class can be void, so if I make them optional, it fixes things:
```
class GraphQLError {
  message: string;
  locations?: Array<{ line: number, column: number }> | void;
  path?: Array<string | number> | void;
  nodes?: Array<ASTNode> | void;
  source?: Source | void;
  positions?: Array<number> | void;
}
```

Would love to fix this :) I'm doubtful they're going to change anything upstream

**Comment by TonyPythoneer** [comment]
I hope keeping the current change. This doubt leaves for next PR.

**Comment by TonyPythoneer** [comment]
Hi @andy-ms @vvakame 
Would you merge this? thx.

**Pull request merged**